Opengroup Poller Reaction Conflict Resolution + Fixes

yougotwill opened the pull request:
- Created a cache for opengroup reactions, which we use to create an "optimistic" state for message reactions in an opengroup (this compares local state vs server state and resolves).
- Added more information about the current number of reactors for an opengroup message when viewing the Reaction List Modal
- Fixed reaction Notifications for 1-1 conversations.
- In opengroups, when sending a reaction or clearing a reaction we update the local state triggering a UI update when the message is sent and if anything failed it will be resolved by the poll update since we overwrite all local state of reactions in open groups. This improves the perceived responsiveness of reacting to a message or moderators clearing a reaction.

yougotwill committed:
feat: created a cache for opengroups reactions, we can add and update them

yougotwill committed:
fix: added current emoji to reaction list reactor description

yougotwill committed:
feat: open group messages are now procesed via the cache

cached entries are now added or removed based on the "optimistic" state that we want

yougotwill committed:
feat: moderator clear all reactions behaviour now uses the cache

yougotwill committed:
fix: repaired reaction notifications for 1-1s

Bilb commented on ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts:
```diff
@@ -0,0 +1,137 @@
+            },
+          });
+          if (reactionMatches?.length) {
+            for (const reactionMatch of reactionMatches) {
```
same coment a s above. the `if` is useless I suppose

yougotwill replied:
Fixed

Bilb commented on ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts:
```diff
@@ -0,0 +1,137 @@
+              }
+            }
+
+            const removedMatches = remove(sogsMutationCache, ...roomMatches);
```
I am not too sure if this is the right spot for this. Shouldn't you do this outside of the main `for..of`

Bilb replied:
or you didn't take into account that `remove` mutates the array?

yougotwill replied:
Good catch, I had put it in the wrong location by mistake. I'm aware it mutates the array 👍

Bilb commented on ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts:
```diff
@@ -0,0 +1,137 @@
+  const updatedReactions = message.reactions;
+  const roomMatches: Array<SogsV3Mutation> = filter(sogsMutationCache, { server, room });
+
+  if (roomMatches?.length) {
+    for (const roomMatch of roomMatches) {
```
if the type complains to be undefined, you can always have it like that

```ts
  const roomMatches: Array<SogsV3Mutation> = filter(sogsMutationCache, { server, room }) || [];
```

Bilb replied:
but looking at your `  const roomMatches: Array<SogsV3Mutation> ` I would say it's all good already

yougotwill replied:
Yeah good point, I think I had that check as I was writing the code but didn't refactor after.

Bilb commented on ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts:
```diff
@@ -0,0 +1,137 @@
+          }
+        }
+      }
+    }
+  }
+
+  message.reactions = updatedReactions;
+  await handleOpenGroupMessageReactions(message.reactions, message.id);
+}
```
just overall comment, but I feel like that `processMessagesUsingCache` could use some testing

yougotwill replied:
I can write up a test file to start with but it will be expanded on in the future when we start using it for other open group behaviours right?

Bilb replied:
yep

yougotwill replied:
done

yougotwill committed:
fix: speed up reaction UI update for opengroups

yougotwill commented on ts/components/dialog/ReactListModal.tsx:
```diff
@@ -274,9 +284,7 @@ export const ReactListModal = (props: Props): ReactElement => {
   const dispatch = useDispatch();
 
   const { convoId, isPublic } = msgProps;
-
-  const convo = getConversationController().get(convoId);
-  const weAreModerator = convo.getConversationModelProps().weAreModerator;
+  const weAreModerator = useWeAreModerator(convoId);
```
Good catch!

yougotwill commented on ts/util/reactions.ts:
```diff
@@ -124,7 +124,12 @@ export const sendMessageReaction = async (messageId: string, emoji: string) => {
       emoji,
       'reaction for message',
-      id
+      id,
+      found.get('isPublic')
+        ? `on ${conversationModel.toOpenGroupV2().serverUrl}/${
+            conversationModel.toOpenGroupV2().roomId
+          }`
+        : ''
```
Should we update it to use `serverUrl/roomId` ?

Bilb replied:
yep works for me

yougotwill replied:
Updated

yougotwill commented on ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts:
```diff
@@ -0,0 +1,137 @@
+  } else {
+    const entryIndex = findIndex(sogsMutationCache, entry);
+    if (entryIndex >= 0) {
+      const updatedEntry = entry;
+      updatedEntry.seqno = seqno;
+      sogsMutationCache[entryIndex] = updatedEntry;
+      window.log.info('SOGS Mutation Cache: Entry updated!', updatedEntry);
```
Good catch!

yougotwill committed:
fix: pr review fixes

yougotwill committed:
test: added tests for adding and updating sogs cache entries

updated idForLogging for opengroups to be more verbose, updated reaction method calls to use exported Reactions object

yougotwill committed:
fix: dont log removing entries from cache when there are none

yougotwill committed:
test: finished writing tests for sogs mutation cache

refactored processMessagesWithCachce function

yougotwill commented on ts/test/session/unit/receiver/opengroup/deduplicate_test.ts:
```diff
@@ -93,20 +93,6 @@ describe('filterDuplicatesFromDbAndIncoming', () => {
       expect(filtered.length).to.be.eq(1);
       expect(filtered[0]).to.be.deep.eq(msg1);
     });
-
```
This was a duplicate test that I cleaned up

Bilb commented on ts/test/session/unit/receiver/opengroup/deduplicate_test.ts:
```diff
@@ -93,20 +93,6 @@ describe('filterDuplicatesFromDbAndIncoming', () => {
       expect(filtered[0]).to.be.deep.eq(msg1);
     });
-
-    it('three duplicates in the same poll', async () => {
-      const msg1 = TestUtils.generateOpenGroupMessageV2();
-      const msg2 = TestUtils.generateOpenGroupMessageV2();
-
-      const msg3 = TestUtils.generateOpenGroupMessageV2();
-      msg2.sentTimestamp = msg1.sentTimestamp;
-      msg2.sender = msg1.sender;
-      msg3.sentTimestamp = msg1.sentTimestamp;
-      msg3.sender = msg1.sender;
-      const filtered = await filterDuplicatesFromDbAndIncoming([msg1, msg2, msg3]);
-      expect(filtered.length).to.be.eq(1);
-      expect(filtered[0]).to.be.deep.eq(msg1);
-    });
```
what happened here?

yougotwill replied:
I made a comment but GitHub messed it up for some reason. It's a duplicate of the one above.

yougotwill committed:
fix: make sure to clear a reaction is the count is 0 on an opengroup
